### ISSUE TITLE
500 error on non-existent resources

### ISSUE THREAD

**Issue by carterpage** [skyscreamer/yoga#52]
If you go to http://yoga-demo-springmvc.cloudfoundry.com/user/100.json, you get a 500 error.  It should return something more user friendly.

exception

org.springframework.web.util.NestedServletException: Request processing failed; nested exception is org.hibernate.ObjectNotFoundException: No row with the given identifier exists: [org.skyscreamer.yoga.demo.model.User#100]
    org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:656)
    org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:549)
    javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
    javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
    org.springframework.orm.hibernate3.support.OpenSessionInViewFilter.doFilterInternal(OpenSessionInViewFilter.java:198)
    org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:76)

root cause

org.hibernate.ObjectNotFoundException: No row with the given identifier exists: [org.skyscreamer.yoga.demo.model.User#100]
    org.hibernate.impl.SessionFactoryImpl$2.handleEntityNotFound(SessionFactoryImpl.java:447)
    org.hibernate.proxy.AbstractLazyInitializer.checkTargetState(AbstractLazyInitializer.java:189)
    org.hibernate.proxy.AbstractLazyInitializer.initialize(AbstractLazyInitializer.java:178)
    org.hibernate.proxy.AbstractLazyInitializer.getImplementation(AbstractLazyInitializer.java:215)
    org.hibernate.Hibernate.getClass(Hibernate.java:381)
    org.skyscreamer.yoga.demo.mapper.HibernateResultTraverser.getClass(HibernateResultTraverser.java:12)
    org.skyscreamer.yoga.mapper.ResultTraverser.traverse(ResultTraverser.java:34)
    org.skyscreamer.yoga.springmvc.view.JsonSelectorView.getSingleResult(JsonSelectorView.java:45)
    org.skyscreamer.yoga.springmvc.view.JsonSelectorView.render(JsonSelectorView.java:32)
    org.skyscreamer.yoga.springmvc.view.AbstractYogaView.render(AbstractYogaView.java:41)
    org.springframework.web.servlet.DispatcherServlet.render(DispatcherServlet.java:1047)
    org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:817)
    org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:719)
    org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:644)
    org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:549)
    javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
    javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
    org.springframework.orm.hibernate3.support.OpenSessionInViewFilter.doFilterInternal(OpenSessionInViewFilter.java:198)
    org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:76)

**Comment by sduskis** [comment]
What should the response be?  404s?  We'll have to explore some sort of ObjectNotFoundException handler in both SpringMVC and RESTeasy

**Comment by carterpage** [comment]
Trying to use @ExceptionHandler(ObjectNotFoundException.class) in the AbstractController, but for some odd reason, the dao does not throw an exception, rather it returns a proxied POJO that contains information regarding a com.sun.jdi.InvocationException.  The exception doesn't get called until here, when Hibernate.getClass gets called in the ViewResolver:

Thread [2124414527@qtp-1783079124-0](Suspended)  
    HibernateResultTraverser.getClass(Object) line: 12  
    HibernateResultTraverser(ResultTraverser).traverse(Object, Selector, HierarchicalModel) line: 34  
    JsonSelectorView.getSingleResult(Object, Selector) line: 45 
    JsonSelectorView.render(OutputStream, Selector, Object) line: 32  
    JsonSelectorView(AbstractYogaView).render(Map<String,?>, HttpServletRequest, HttpServletResponse) line: 41  
    DispatcherServlet.render(ModelAndView, HttpServletRequest, HttpServletResponse) line: 1047  
    DispatcherServlet.doDispatch(HttpServletRequest, HttpServletResponse) line: 817 
    DispatcherServlet.doService(HttpServletRequest, HttpServletResponse) line: 719  
    DispatcherServlet(FrameworkServlet).processRequest(HttpServletRequest, HttpServletResponse) line: 644  
    DispatcherServlet(FrameworkServlet).doGet(HttpServletRequest, HttpServletResponse) line: 549  
    DispatcherServlet(HttpServlet).service(HttpServletRequest, HttpServletResponse) line: 707  
    DispatcherServlet(HttpServlet).service(ServletRequest, ServletResponse) line: 820  
    ServletHolder.handle(ServletRequest, ServletResponse) line: 511 
    ServletHandler$CachedChain.doFilter(ServletRequest, ServletResponse) line: 1166 
    OpenSessionInViewFilter.doFilterInternal(HttpServletRequest, HttpServletResponse, FilterChain) line: 198  
    OpenSessionInViewFilter(OncePerRequestFilter).doFilter(ServletRequest, ServletResponse, FilterChain) line: 76  
    ServletHandler$CachedChain.doFilter(ServletRequest, ServletResponse) line: 1157 
    ServletHandler.handle(String, HttpServletRequest, HttpServletResponse, int) line: 388  
    SecurityHandler.handle(String, HttpServletRequest, HttpServletResponse, int) line: 216  
    SessionHandler.handle(String, HttpServletRequest, HttpServletResponse, int) line: 182  
    WebAppContext(ContextHandler).handle(String, HttpServletRequest, HttpServletResponse, int) line: 765  
    WebAppContext.handle(String, HttpServletRequest, HttpServletResponse, int) line: 418  
    Server(HandlerWrapper).handle(String, HttpServletRequest, HttpServletResponse, int) line: 152  
    Server.handle(HttpConnection) line: 326 
    HttpConnection.handleRequest() line: 542  
    HttpConnection$RequestHandler.headerComplete() line: 926  
    HttpParser.parseNext() line: 549  
    HttpParser.parseAvailable() line: 212  
    HttpConnection.handle() line: 404  
    SocketConnector$Connection.run() line: 228  
    QueuedThreadPool$PoolThread.run() line: 582 

wtf?

**Comment by carterpage** [comment]
My problem was with Hibernate 3's lazy-initialized entities.  I called Hibernate.getClass(entity) to force a load in the dao.  (Other alternatives would be to replace load with get and throw exceptions, or to override the default lazy initialization from the hibernate config.  I think the route I chose is the least intrusive.)

For more information on the lazy init: http://www.sleberknight.com/blog/sleberkn/entry/20051023
